File: src/lib/esplora.ts
import { ESPLORA_BASE, type NetworkName } from '../config'
import type { Utxo } from './bitcoin'

function base(net: NetworkName): string {
  return ESPLORA_BASE[net]
}

export interface AddressStats {
  confirmed: number // sats
  pending: number // sats (unconfirmed mempool delta, can be negative)
}

export async function fetchBalance(net: NetworkName, address: string): Promise<AddressStats> {
  const res = await fetch(`${base(net)}/address/${address}`)
  if (!res.ok) throw new Error(`Couldn’t fetch balance (${res.status})`)
  const j = await res.json()
  const confirmed = j.chain_stats.funded_txo_sum - j.chain_stats.spent_txo_sum
  const pending = j.mempool_stats.funded_txo_sum - j.mempool_stats.spent_txo_sum
  return { confirmed, pending }
}

export async function fetchUtxos(net: NetworkName, address: string): Promise<Utxo[]> {
  const res = await fetch(`${base(net)}/address/${address}/utxo`)
  if (!res.ok) throw new Error(`Couldn’t fetch coins (${res.status})`)
  const j = await res.json()
  return (j as any[]).map((u) => ({ txid: u.txid, vout: u.vout, value: u.value }))
}

/** Recommended fee in sat/vByte (half-hour target), with sane fallbacks. */
export async function fetchFeeRate(net: NetworkName): Promise<number> {
  try {
    const res = await fetch(`${base(net)}/v1/fees/recommended`)
    if (res.ok) {
      const j = await res.json()
      const rate = j.halfHourFee ?? j.hourFee ?? j.economyFee
      if (typeof rate === 'number' && rate > 0) return rate
    }
  } catch {
    /* fall through to default */
  }
  return net === 'mainnet' ? 8 : 1
}

/** Current BTC price in USD from mempool.space, or null if unavailable. */
export async function fetchPriceUsd(): Promise<number | null> {
  try {
    const res = await fetch('https://mempool.space/api/v1/prices')
    if (!res.ok) return null
    const j = await res.json()
    return typeof j.USD === 'number' && j.USD > 0 ? j.USD : null
  } catch {
    return null
  }
}

export interface AddressStatsFull {
  confirmed: number
  pending: number
  used: boolean // has any on-chain or mempool history
}

/** Address balance + whether it's ever been used (for HD gap-limit scanning). */
export async function fetchAddressStats(net: NetworkName, address: string): Promise<AddressStatsFull> {
  const res = await fetch(`${base(net)}/address/${address}`)
  if (!res.ok) throw new Error(`Couldn’t fetch address (${res.status})`)
  const j = await res.json()
  const confirmed = j.chain_stats.funded_txo_sum - j.chain_stats.spent_txo_sum
  const pending = j.mempool_stats.funded_txo_sum - j.mempool_stats.spent_txo_sum
  const used = j.chain_stats.tx_count + j.mempool_stats.tx_count > 0
  return { confirmed, pending, used }
}

/** Recent transactions across ALL the wallet's addresses, deduped, with wallet-level deltas. */
export async function fetchWalletHistory(net: NetworkName, addresses: string[]): Promise<TxSummary[]> {
  if (addresses.length === 0) return []
  const ours = new Set(addresses)
  const byTxid = new Map<string, any>()
  for (const addr of addresses) {
    try {
      const res = await fetch(`${base(net)}/address/${addr}/txs`)
      if (!res.ok) continue
      for (const tx of (await res.json()) as any[]) if (!byTxid.has(tx.txid)) byTxid.set(tx.txid, tx)
    } catch {
      /* skip this address on error */
    }
  }
  const out: TxSummary[] = []
  for (const tx of byTxid.values()) {
    let delta = 0
    for (const o of tx.vout) if (ours.has(o.scriptpubkey_address)) delta += o.value
    for (const i of tx.vin) if (i.prevout && ours.has(i.prevout.scriptpubkey_address)) delta -= i.prevout.value
    out.push({
      txid: tx.txid,
      delta,
      confirmed: tx.status?.confirmed ?? false,
      blockTime: tx.status?.block_time,
    })
  }
  out.sort((a, b) => {
    if (a.confirmed !== b.confirmed) return a.confirmed ? 1 : -1 // pending first
    return (b.blockTime ?? 0) - (a.blockTime ?? 0)
  })
  return out
}

export async function broadcastTx(net: NetworkName, hex: string): Promise<string> {
  const res = await fetch(`${base(net)}/tx`, {
    method: 'POST',
    headers: { 'Content-Type': 'text/plain' },
    body: hex,
  })
  const text = (await res.text()).trim()
  if (!res.ok) throw new Error(text || `Broadcast failed (${res.status})`)
  return text // txid
}

export interface TxSummary {
  txid: string
  delta: number // net effect on the address: + received, - sent (sats)
  confirmed: boolean
  blockTime?: number
}

export async function fetchHistory(net: NetworkName, address: string): Promise<TxSummary[]> {
  const res = await fetch(`${base(net)}/address/${address}/txs`)
  if (!res.ok) throw new Error(`Couldn’t fetch history (${res.status})`)
  const txs = await res.json()
  return (txs as any[]).map((tx) => {
    let delta = 0
    for (const o of tx.vout) if (o.scriptpubkey_address === address) delta += o.value
    for (const i of tx.vin) if (i.prevout?.scriptpubkey_address === address) delta -= i.prevout.value
    return {
      txid: tx.txid,
      delta,
      confirmed: tx.status?.confirmed ?? false,
      blockTime: tx.status?.block_time,
    }
  })
}

File: android/app/src/main/java/app/wolit/net/Esplora.java
package app.wolit.net;

import app.wolit.Config;
import app.wolit.bitcoin.Utxo;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Chain data via the public mempool.space / Esplora API. Mirrors src/lib/esplora.ts.
 */
public final class Esplora {
    private Esplora() {}

    public static final class AddressStats {
        public final long confirmed; // sats
        public final long pending;   // sats (unconfirmed mempool delta, may be negative)
        public AddressStats(long confirmed, long pending) {
            this.confirmed = confirmed;
            this.pending = pending;
        }
    }

    public static final class TxSummary {
        public final String txid;
        public final long delta;       // + received, - sent (sats)
        public final boolean confirmed;
        public final long blockTime;   // unix seconds, 0 if pending
        public TxSummary(String txid, long delta, boolean confirmed, long blockTime) {
            this.txid = txid;
            this.delta = delta;
            this.confirmed = confirmed;
            this.blockTime = blockTime;
        }
    }

    private static String base() {
        return Config.esploraBase();
    }

    public static AddressStats fetchBalance(String address) throws IOException {
        Http.Response r = Http.get(base() + "/address/" + address, null);
        if (!r.ok()) throw new IOException("Couldn’t fetch balance (" + r.code + ")");
        try {
            JSONObject j = new JSONObject(r.body);
            JSONObject chain = j.getJSONObject("chain_stats");
            JSONObject mem = j.getJSONObject("mempool_stats");
            long confirmed = chain.getLong("funded_txo_sum") - chain.getLong("spent_txo_sum");
            long pending = mem.getLong("funded_txo_sum") - mem.getLong("spent_txo_sum");
            return new AddressStats(confirmed, pending);
        } catch (Exception e) {
            throw new IOException("Unexpected balance response", e);
        }
    }

    public static List<Utxo> fetchUtxos(String address) throws IOException {
        Http.Response r = Http.get(base() + "/address/" + address + "/utxo", null);
        if (!r.ok()) throw new IOException("Couldn’t fetch coins (" + r.code + ")");
        try {
            JSONArray arr = new JSONArray(r.body);
            List<Utxo> out = new ArrayList<>(arr.length());
            for (int i = 0; i < arr.length(); i++) {
                JSONObject u = arr.getJSONObject(i);
                out.add(new Utxo(u.getString("txid"), u.getInt("vout"), u.getLong("value")));
            }
            return out;
        } catch (Exception e) {
            throw new IOException("Unexpected UTXO response", e);
        }
    }

    /** Recommended fee in sat/vByte (half-hour target), with sane fallbacks. */
    public static double fetchFeeRate() {
        try {
            Http.Response r = Http.get(base() + "/v1/fees/recommended", null);
            if (r.ok()) {
                JSONObject j = new JSONObject(r.body);
                double rate = j.optDouble("halfHourFee",
                        j.optDouble("hourFee", j.optDouble("economyFee", 0)));
                if (rate > 0) return rate;
            }
        } catch (Exception ignored) {
            // fall through to default
        }
        return Config.isMainnet() ? 8 : 1;
    }

    /** Current BTC price in USD, or null if unavailable. */
    public static Double fetchPriceUsd() {
        try {
            Http.Response r = Http.get("https://mempool.space/api/v1/prices", null);
            if (!r.ok()) return null;
            JSONObject j = new JSONObject(r.body);
            double usd = j.optDouble("USD", 0);
            return usd > 0 ? usd : null;
        } catch (Exception e) {
            return null;
        }
    }

    public static String broadcast(String hex) throws IOException {
        Http.Response r = Http.request("POST", base() + "/tx", null, "text/plain",
                hex.getBytes(java.nio.charset.StandardCharsets.UTF_8));
        String text = r.body == null ? "" : r.body.trim();
        if (!r.ok()) throw new IOException(text.isEmpty() ? "Broadcast failed (" + r.code + ")" : text);
        return text; // txid
    }

    public static List<TxSummary> fetchHistory(String address) throws IOException {
        Http.Response r = Http.get(base() + "/address/" + address + "/txs", null);
        if (!r.ok()) throw new IOException("Couldn’t fetch history (" + r.code + ")");
        try {
            JSONArray txs = new JSONArray(r.body);
            List<TxSummary> out = new ArrayList<>(txs.length());
            for (int i = 0; i < txs.length(); i++) {
                JSONObject tx = txs.getJSONObject(i);
                long delta = 0;
                JSONArray vout = tx.getJSONArray("vout");
                for (int k = 0; k < vout.length(); k++) {
                    JSONObject o = vout.getJSONObject(k);
                    if (address.equals(o.optString("scriptpubkey_address", null))) {
                        delta += o.getLong("value");
                    }
                }
                JSONArray vin = tx.getJSONArray("vin");
                for (int k = 0; k < vin.length(); k++) {
                    JSONObject in = vin.getJSONObject(k);
                    JSONObject prev = in.optJSONObject("prevout");
                    if (prev != null && address.equals(prev.optString("scriptpubkey_address", null))) {
                        delta -= prev.getLong("value");
                    }
                }
                JSONObject status = tx.optJSONObject("status");
                boolean confirmed = status != null && status.optBoolean("confirmed", false);
                long blockTime = status != null ? status.optLong("block_time", 0) : 0;
                out.add(new TxSummary(tx.getString("txid"), delta, confirmed, blockTime));
            }
            return out;
        } catch (IOException e) {
            throw e;
        } catch (Exception e) {
            throw new IOException("Unexpected history response", e);
        }
    }

    public static List<TxSummary> emptyHistory() {
        return Collections.emptyList();
    }
}
